### PR TITLE
[CSSolver] Make sure generic requirements are checked for typealiases

### PR DESCRIPTION
**Pull request by xedin**
If solver encounters a typealias inside of constrainted extension
make sure to add its requirements to the constraint system, otherwise
it might produce invalid solutions.

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

File: lib/Sema/ConstraintSystem.h [hunk 1]
```diff
@@ -2159,6 +2159,11 @@ class ConstraintSystem {
                    bool skipProtocolSelfConstraint,
                    ConstraintLocatorBuilder locator,
                    OpenedTypeMap &replacements);
+  void openGenericRequirements(DeclContext *outerDC,
```
Comment (rudkx) [inline lib/Sema/ConstraintSystem.h]: Can you add a comment explaining what this does?

File: lib/Sema/ConstraintSystem.cpp [hunk 2]
```diff
@@ -486,8 +486,25 @@ Type ConstraintSystem::openUnboundGenericType(
...
+  // let's add all of the generic requirements to the constraint system to
+  // make sure that it's something we can use.
+  if (auto *NAT = dyn_cast<NameAliasType>(type.getPointer())) {
+    auto *decl = NAT->getDecl();
+    auto *extension = dyn_cast<ExtensionDecl>(decl->getDeclContext());
+    if (extension && extension->isConstrainedExtension()) {
+      auto substitutions = NAT->getSubstitutionMap();
+      if (auto *signature = substitutions.getGenericSignature())
+        openGenericRequirements(
+            decl->getDeclContext(), signature,
```
Comment (rudkx) [inline lib/Sema/ConstraintSystem.cpp]: Nit: You could just pass `extension` here.

**Review by rudkx — COMMENTED** [review]
Otherwise LGTM.

**Review by DougGregor — APPROVED** [review]
LGTM, but please address @rudkx 's comments

**Comment by xedin** [comment]
Thanks, @DougGregor! I was just waiting for you to take a look before I make these changes...

File: lib/Sema/ConstraintSystem.h (hunk 1 above)
Comment (xedin) [reply lib/Sema/ConstraintSystem.h]: Done!

File: lib/Sema/ConstraintSystem.cpp (hunk 2 above)
Comment (xedin) [reply lib/Sema/ConstraintSystem.cpp]: Done!

**Comment by xedin** [comment]
Apparently contextual substitutions on parent where needed as well, I've added additional tests to cover that.

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
Looks like all of the tests now pass, merging.

**Pull request merged**